pkg/ovs/openflow/ofctrl_packetin.go

```diff
@@ -40,15 +52,31 @@ func GetTCPHeaderData(ipPkt util.Message) (tcpSrcPort, tcpDstPort uint16, tcpSeq
 		tcpBytes, err = typedIPPkt.Data.(*util.Buffer).MarshalBinary()
 	}
 	if err != nil {
-		return 0, 0, 0, 0, 0, err
+		return nil, err
 	}
-	tcpIn := new(protocol.TCP)
-	err = tcpIn.UnmarshalBinary(tcpBytes)
+	tcpPkt = new(protocol.TCP)
+	err = tcpPkt.UnmarshalBinary(tcpBytes)
 	if err != nil {
-		return 0, 0, 0, 0, 0, err
+		return nil, err
 	}
 
-	return tcpIn.PortSrc, tcpIn.PortDst, tcpIn.SeqNum, tcpIn.AckNum, tcpIn.Code, nil
+	return tcpPkt, nil
+}
+
+func GetTCPDNSData(tcpPkt *protocol.TCP) (data []byte, err error) {
+	// TCP.HdrLen is 4-octet unit indicating the length of TCP header including options.
+	tcpOptionsLen := (tcpPkt.HdrLen - tcpStandardHdrLen) * 4
+	// Move two more octet.
+	// From RFC 7766:
+	// DNS clients and servers SHOULD pass the two-octet length field, and
+	// the message described by that length field, to the TCP layer at the
+	// same time (e.g., in a single "write" system call) to make it more
+	// likely that all the data will be transmitted in a single TCP segment.
+	dnsDataStartIdx := tcpOptionsLen + 2
+	if int(dnsDataStartIdx) > len(tcpPkt.Data) {
+		return nil, fmt.Errorf("no DNS data in TCP data")
+	}
+	return tcpPkt.Data[dnsDataStartIdx:], nil
```

Comment:
To be more robust, I feel it should check if the length of data is shorter than the length field.
If yes, it means the data cannot be contained in a single packet and there should be a following packet.
I guess the first packet should at least contain all answers and we don't need to care the following packets in most cases but I'm not sure whether the unpack function can partially unpack the data without throwing errors. I think we should at least log an info here to indicate this happens and it will be helpful for troubleshooting.

Reply:
Additional check added. 
And I did a quick check: `unpack` couldn't partially unpack DNS. There are small chances that the DNS response is over the MTU, so maybe we could put "handle fragmented TCP DNS response" in a future iteration.